java-checks/src/main/java/org/sonar/java/checks/CloseResourceCheck.java

```diff
@@ -0,0 +1,400 @@
+
+      visitBlock(tree.block(), currentES, endES);
+
+      for (CatchTree catchTree : tree.catches()) {
+        visitBlock(catchTree.block(), new ExecutionState(currentES), endES);
+      }
+
+      if (tree.finallyBlock() != null) {
+        visitBlock(tree.finallyBlock(), new ExecutionState(currentES), endES, true);
+      }
+
+      executionState = endES;
+    }
+
+    private void visitBlock(BlockTree block, ExecutionState startES, ExecutionState endEs) {
+      visitBlock(block, startES, endEs, false);
+    }
+
+    private void visitBlock(BlockTree block, ExecutionState startES, ExecutionState endEs, boolean forceState) {
+      ExecutionState currentES = new ExecutionState(startES);
+      executionState = currentES;
+      block.accept(this);
+      endEs.merge(currentES, forceState);
+    }
+
+    @Override
+    public void visitIfStatement(IfStatementTree tree) {
+      tree.condition().accept(this);
+
+      ExecutionState currentES = executionState;
+      ExecutionState thenES = new ExecutionState(currentES);
+      ExecutionState elseES = new ExecutionState(currentES);
+
+      executionState = thenES;
+      tree.thenStatement().accept(this);
+
+      if (tree.elseStatement() != null) {
+        executionState = elseES;
+        tree.elseStatement().accept(this);
+      }
+      executionState = currentES.merge(thenES, elseES);
+    }
+
+    private Set<Symbol> extractCloseableSymbols(List<VariableTree> variableTrees) {
+      Set<Symbol> symbols = Sets.newHashSet();
+      for (VariableTree variableTree : variableTrees) {
+        Symbol symbol = variableTree.symbol();
+        if (symbol.type().isSubtypeOf(JAVA_IO_CLOSEABLE)) {
+          symbols.add(symbol);
+        }
+      }
+      return symbols;
+    }
+
+    private static class ExecutionState {
+      private Set<Symbol> excludedCloseables;
+      private Map<Symbol, CloseableOccurence> closeableOccurenceBySymbol = Maps.newHashMap();
+      private Set<Tree> unclosedCloseableReferences = Sets.newHashSet();
+
+      ExecutionState(Set<Symbol> excludedCloseables) {
+        this.excludedCloseables = excludedCloseables;
+      }
+
+      public ExecutionState(ExecutionState baseES) {
+        this.excludedCloseables = Sets.newHashSet(baseES.excludedCloseables);
+        for (Entry<Symbol, CloseableOccurence> entry : baseES.closeableOccurenceBySymbol.entrySet()) {
+          this.closeableOccurenceBySymbol.put(entry.getKey(), new CloseableOccurence(entry.getValue()));
+        }
+        // the unclosed closeable set is let empty
+      }
+
+      public ExecutionState merge(ExecutionState es, boolean forceNewState) {
+        return this.merge(es, null, forceNewState);
+      }
+
+      public ExecutionState merge(ExecutionState es1, @Nullable ExecutionState es2) {
+        return this.merge(es1, es2, false);
+      }
+
+      public ExecutionState merge(ExecutionState es1, @Nullable ExecutionState es2, boolean forceFirstState) {
```

Comment:
This third parameter should probably not exist and its functionality be covered in another method